shape_user_object parallel failure fix

roystgnr opened the pull request:
This fixes the userobjects/shape_element_user_object.shape_element_user_object test for me on 13+ processors (and hopefully for Civet too - it was the nightly tests from #9430 that caught the overzealous error.

This also re-enables that suite of tests when running on DistributedMesh, which should help avoid future regressions there or in #1500.

roystgnr committed:
Add const SystemBase::system() accessor

roystgnr committed:
Add const FEProblemBase::getNonlinearSystemBase()

roystgnr committed:
Don't error on n_local_dofs==0 corner case

This fixes the
userobjects/shape_element_user_object.shape_element_user_object test
for me on 13+ processors (and hopefully for Civet too - it was the
nightly tests from #9430 that caught the overzealous error.

roystgnr committed:
shape_element_user_object works distributed too

And we've now changed the tests which use non-DistributedMesh
compatible code so they build a ReplicatedMesh regardless.

So run the tests.  This must be why I wasn't noticing the test
failures here on my --enable-parmesh libMesh build.

This should help avoid future regressions in this test or in #1500.

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/110520/) on e947306 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 76affa44fb25adea03ebc27324b330d28348c6a0`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9723/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9723/style.patch | git apply -v`

roystgnr committed:
clang-format fix

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/110546/) on 80110a1 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/9723/site)

dschwen commented on test/src/userobjects/TestShapeElementUserObject.C:
```diff
@@ -74,12 +76,21 @@ TestShapeElementUserObject::executeJacobian(unsigned int jvar)
+  // check if executeJacobian was called for each variable on at least
+  // one MPI process.  We'd like to make sure that executeJacobian was
+  // called on *every* process which owns degrees of freedom, but I
+  // don't think we
```
that train of thought looks unfinished

roystgnr replied:
Ha!  I was starting to write an explanation of why I couldn't make this test as robust as it should be, then I suddenly had an idea about how to make this test as robust as it should be, then I figured out how to actually implement that idea, did that, tested it... and by this time had completely forgotten that I'd started writing a comment in there too.

It's fixup'ed now.

roystgnr committed:
Build ReplicatedMesh on Nonlocal kernel tests

I'd still like to *run* these tests when I've built a
DistributedMesh-default libMesh, but nonlocal kernels aren't yet
compatible with DistributedMesh
